File: Solutions/Neetcode/Stack/3-EvaluateReversePolishNotation/solution.ts
function evalRPN(tokens: string[]): number {
  const numStack: number[] = [];

  for (let token of tokens) {
    if (!isNaN(Number(token))) {
      numStack.push(Number(token));
      continue;
    }

    const num2 = numStack.pop()!;
    const num1 = numStack.pop()!;
    if (token === "+") {
      numStack.push(Number(num1) + Number(num2));
    } else if (token === "-") {
      numStack.push(num1 - num2);
    } else if (token === "*") {
      numStack.push(num1 * num2);
    } else if (token === "/") {
      numStack.push(Math.trunc(num1 / num2));
    }
  }

  return numStack[0];
}

File: Solutions/Neetcode/Stack/3-EvaluateReversePolishNotation/solution.py
from typing import List
from math import trunc


class Solution:
    def evalRPN(self, tokens: List[str]) -> int:
        stack = []

        for t in tokens:
            if t == "+":
                num2 = stack.pop()
                num1 = stack.pop()
                stack.append(num1 + num2)
            elif t == "-":
                num2 = stack.pop()
                num1 = stack.pop()
                stack.append(num1 - num2)
            elif t == "/":
                num2 = stack.pop()
                num1 = stack.pop()
                stack.append(trunc(num1 / num2))

            elif t == "*":
                num2 = stack.pop()
                num1 = stack.pop()
                stack.append(num1 * num2)
            else:
                stack.append(int(t))

        return stack[0]
